src/CloudNative.CloudEvents.SystemTextJson/JsonEventFormatter.cs

```diff
@@ -0,0 +1,489 @@
+                    JsonValueKind.Null => null,
+                    // Note: this will fail if the value isn't an integer, or is out of range for Int32.
+                    JsonValueKind.Number => CloudEventAttributeType.Integer.Format(value.GetInt32()),
+                    _ => throw new ArgumentException($"Invalid token type '{value.ValueKind}' for CloudEvent attribute")
+                };
+                if (attributeValue is null)
+                {
+                    continue;
+                }
+                // Note: we *could* infer an extension type of integer and Boolean, but not other extension types.
+                // (We don't want to assume that everything that looks like a timestamp is a timestamp, etc.)
+                // Stick to strings for consistency.
+                cloudEvent.SetAttributeFromString(name, attributeValue);
+            }
+        }
+
+        private void ValidateTokenTypeForAttribute(CloudEventAttribute attribute, JsonValueKind valueKind)
+        {
+            // We can't validate unknown attributes, don't check for extension attributes,
+            // and null values will be ignored anyway.
+            if (attribute is null || attribute.IsExtension || valueKind == JsonValueKind.Null)
+            {
+                return;
+            }
+
+            // This is deliberately written so that if a new attribute type is added without this being updated, we "fail valid".
+            // (That should only happen in major versions anyway, but it's worth being somewhat forgiving here.)
+            // TODO: Can we avoid hard-coding the strings here? We could potentially introduce an enum for attribute types.
+            var valid = attribute.Type.Name switch
+            {
+                "Binary" => valueKind == JsonValueKind.String,
+                "Boolean" => valueKind == JsonValueKind.True || valueKind == JsonValueKind.False,
+                "Integer" => valueKind == JsonValueKind.Number,
+                "String" => valueKind == JsonValueKind.String,
+                "Timestamp" => valueKind == JsonValueKind.String,
+                "URI" => valueKind == JsonValueKind.String,
+                "URI-Reference" => valueKind == JsonValueKind.String,
+                _ => true
+            };
+            if (!valid)
+            {
+                throw new ArgumentException($"Invalid token type '{valueKind}' for CloudEvent attribute '{attribute.Name}' with type '{attribute.Type}'");
+            }
+        }
+
+        private void PopulateDataFromStructuredEvent(CloudEvent cloudEvent, JsonDocument document)
+        {
+            // Fetch data and data_base64 tokens, and treat null as missing.
+            document.RootElement.TryGetProperty(DataPropertyName, out var dataElement);
+            if (dataElement is JsonElement { ValueKind: JsonValueKind.Null })
+            {
+                dataElement = new JsonElement();
```

Comment:
Hmm, why do we create a new element if it is a Null value kind?

Reply:
So that it's equivalent to not being present (i.e. as if TryGetProperty had returned false). (In Json.NET we just use a null reference, but JsonElement is a struct.)

Reply:
interesting.. I wouldn't have expected that we would mutate the document in this method.

Reply:
We're not - we're just changing the local variable to have a different value. That doesn't mutate the document at all. It's just saying "treat a null value as if it weren't present at all".

Reply:
But why do that? it doesn't appear to be used other than in doing the Undefined check.

Reply:
If we did that, we'd basically need the opposite code, to coalesce undefined to null. An alternative would be to extract two `bool` variables, e.g.

```csharp
bool gotData = dataElement.ValueKind != JsonValueKind.Null && dataElement.ValueKind != JsonValueKind.Undefined;
// Ditto for base64
```

Would you prefer that? Happy to do so.

(I note that the pattern matching isn't required here - that's left over from Json.NET.)

Reply:
I see - we need to handle both Undefined and Null. That was the part I was missing. Thanks!